core/storage/base_model/gae_models_test.py

```diff
@@ -357,6 +358,183 @@ def test_get_new_id_method_does_not_fail_with_bad_names(self):
+
+    def test_put(self):
+        with self.assertRaisesRegexp(
+            NotImplementedError, 'Use put_for_human or put_for_bot instead'):
+            self.model_instance.put()
+
+    def test_put_for_human(self):
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+        previous_last_updated_by_human = (
+            model_instance.last_updated_by_human)
+
+        model_instance.update_timestamps()
+        self.model_instance.put_for_human()
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+
+        self.assertNotEqual(
+            previous_last_updated_by_human,
+            model_instance.last_updated_by_human)
+
+    def test_put_for_bot(self):
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+        previous_last_updated_by_human = (
+            model_instance.last_updated_by_human)
+
+        model_instance.update_timestamps()
+        self.model_instance.put_for_bot()
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+
+        self.assertEqual(
+            previous_last_updated_by_human,
+            model_instance.last_updated_by_human)
+
+    def test_put_async(self):
+        with self.assertRaisesRegexp(
+            NotImplementedError,
+            'Use put_async_for_human or put_async_for_bot instead'):
+            self.model_instance.put_async()
+
+    def test_put_async_for_human(self):
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+        previous_last_updated_by_human = (
+            model_instance.last_updated_by_human)
+
+        model_instance.update_timestamps()
+        self.model_instance.put_async_for_human()
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+
+        self.assertNotEqual(
+            previous_last_updated_by_human,
+            model_instance.last_updated_by_human)
+
+    def test_put_async_for_bot(self):
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+        previous_last_updated_by_human = (
+            model_instance.last_updated_by_human)
+
+        model_instance.update_timestamps()
+        self.model_instance.put_async_for_bot()
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+
+        self.assertEqual(
+            previous_last_updated_by_human,
+            model_instance.last_updated_by_human)
+
+    def test_put_multi(self):
+        with self.assertRaisesRegexp(
+            NotImplementedError,
+            'Use put_multi_for_human or put_multi_for_bot instead'):
+            TestBaseHumanMaintainedModel.put_multi([])
+
+    def test_put_multi_for_human(self):
+        model_instance = TestBaseHumanMaintainedModel.get_by_id(self.MODEL_ID)
+        previous_last_updated_by_human = (
+            model_instance.last_updated_by_human)
+
+        with self.swap(
+            TestBaseHumanMaintainedModel, 'put_async',
```

Comment:
Could you please post the stack trace when you do not mock the `put_async` here?

Reply:
![Screenshot from 2020-11-26 23-18-36](https://user-images.githubusercontent.com/42460632/100381114-43e32f80-303e-11eb-8608-47334a5d6142.png)

Reply:
@brianrodri Hi, can you look at this, since internally ndb `put_multi*` methods call `put_async` method I think we would need some redesign in here because now the `put_multi*` methods do not work.

Reply:
You should be able to overcome this by using super:

```python
super(HumanMaintainedModel, self).put_async(...)
```

Will bypass the check.

Reply:
I am not quite sure where I should be using this. If I use it in put_async of the BaseHumanMaintainedModel then the exception can't be raised and since it is used by the datastore, I am not sure where to place it.

Reply:
You should be able to put it in the test instead of creating a mock model/swap.

Reply:
@brianrodri But how will this help with the larger issue? If I understand it correctly the error above will happen every time that someone call `put_multi` and other multi methods, do we expect everyone to use `super`?

Reply:
@brianrodri ping.

Reply:
Hmm I see now, I'll need to think more deeply about this, I don't see a simple thing.

Reply:
I've spoken with @vojtechjelinek about this and he says he can help redesign the API to fix the issue. I can offer review and guidance but am too busy to handle the whole process :(

Reply:
@brianrodri Yep, thanks, we have discussed this with @rohitkatlaa and he will look into it.

Reply:
@vojtechjelinek @brianrodri I have found a solution and pushed the code. I have create a simple design doc which I have mentioned in the description of this PR.